### PR TITLE
[fix](transaction) Fix concurrent schema change and txn cause dead lock

### PR DESCRIPTION
**Pull request by yujun777**
Concurrent schema change and txn may cause dead lock. An example:
1.  Txn T commit but not publish;
2.  Run schema change or rollup on T's related partition,  add alter replica R;
3.  sc/rollup add a sched txn watermark M; 
4.  Restart fe;
5.  After fe restart,  T's loadedTblIndexes will clear  because it's not save to disk;
6.  T will publish version to all tablet,  including sc/rollup's new alter replica R;
7.  Since R not contains txn data,  so the T will fail. It will then always waitting for R's data;
8.  sc/rollup wait for txn before M to finish,  only after that it will let R copy history data;
9.  Since T's not finished, so sc/rollup will always wait,  so R will nerver copy history data;
10. Txn T and sc/rollup will wait each other forever,   cause dead lock;

Fix:  because sc/rollup will ensure double write after the sched watermark M, so for finish transaction, when checking a alter replica:
1. if txn id is bigger than M,  check it just like a normal replica;
2. otherwise skip check this replica,   the BE will modify history data later.

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

File: be/src/olap/task/engine_publish_version_task.cpp [hunk 1]
```diff
@@ -235,7 +236,19 @@ Status EnginePublishVersionTask::finish() {
             } else {
                 // check if the version exist, if not exist, then set publish failed
                 if (_error_tablet_ids->find(tablet_id) == _error_tablet_ids->end()) {
-                    if (tablet->check_version_exist(version)) {
+                    bool exist_version = tablet->check_version_exist(version);
+                    bool is_converting = false;
+                    if (!exist_version && tablet->tablet_state() == TabletState::TABLET_NOTREADY) {
+                        bool is_mow = tablet->keys_type() == KeysType::UNIQUE_KEYS &&
```
Comment (deardeng) [inline be/src/olap/task/engine_publish_version_task.cpp:242]: add some log?

**Comment by luwei16** [comment]
LGTM

File: fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java [hunk 2]
```diff
@@ -1111,8 +1114,28 @@ public void finishTransaction(long transactionId) throws UserException {
         LOG.info("finish transaction {} successfully, publish result: {}", transactionState, publishResult.name());
     }
 
-    private void checkReplicaContinuousVersionSucc(long tabletId, Replica replica, long version,
-            PublishVersionTask backendPublishTask, Set<Long> errorReplicaIds, List<Replica> tabletSuccReplicas,
+    private boolean isCheckTxnIgnoreAlterReplica(long transactionId, OlapTable table) {
```
[the remark this answers is not in the document]
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/transaction/DatabaseTransactionMgr.java]: updated

File: be/src/olap/task/engine_publish_version_task.cpp (hunk 1 above)
Comment (yujun777) [reply be/src/olap/task/engine_publish_version_task.cpp]: remove be code

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Review by dataroaring — APPROVED** [review]
LGTM

**Review by luwei16 — APPROVED** [review]

**Pull request merged**